### PR TITLE
tools/refengine: Add a ref-engine protocol registry

### PR DESCRIPTION
**Pull request by wking**
And add ref-engine references.  This [cleans up the ref-engine portion of the previous `tools/` implementation][1].  I have tried to touch the ref-engine discovery portion as little as possible, although there's some cleanup work still to do there as well.

I've removed the `tools/` prefix from the `tools/README.md` example, because that README is already inside the `tools` directory.

There are also two more minor commits riding along:

* `tools/hostbasedimagenames`: Add `name` to parsed map
* `Makefile`: Add Go test targets

A bit more motivation for both in their commit messages.

[1]: https://github.com/xiekeyang/oci-discovery/pull/16#discussion_r139475331

### CODE REVIEW

File: tools/newimagespec/newimagespec.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+package v1
+
+import (
+	digest "github.com/opencontainers/go-digest"
+	"github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/xiekeyang/oci-discovery/tools/engine"
+)
+
+type Descriptor struct {
```
Comment (xiekeyang) [inline tools/newimagespec/newimagespec.go]: Can this be defined as below?
```go
type Descriptor struct {
        v1.Descriptor

        // CASEngines specifies a list of CAS engines from which this object
        // and its dependencies MAY be downloaded.
        CASEngines []engine.Config `json:"casEngines,omitempty"`
}
```

Comment (wking) [reply tools/newimagespec/newimagespec.go]: > Can this be defined as below?

Probably.  I'll give that a shot in a few hours.

**Comment by xiekeyang** [comment]
Sorry I should have put comment [1] here.

[1]: https://github.com/xiekeyang/oci-discovery/pull/22#issuecomment-330732882

**Comment by wking** [comment]
> In current code it seems do nothing.
>
> I agree in a net service context help to shutdown some handler gracefully...

That's what it's [doing in this PR][1].  It's also part of [the other ref-engine interface calls][2], because a patricular ref-engine implementation could have slow/blocking code in any of them.

And besides graceful cancels, they're also useful for deadlines/timeouts.

[1]: https://github.com/xiekeyang/oci-discovery/pull/23/files#diff-855940c550098b12c1e2721e8921c216R92
[2]: https://github.com/xiekeyang/oci-discovery/pull/23/files#diff-dd79b1b6e1b0fe7a071f22df18c01029R25

File: tools/hostbasedimagenames/hostbasedimagenames.go [hunk 2]
```diff
@@ -61,5 +61,7 @@ func Parse(name string) (matches map[string]string, err error) {
 		matches["fragment"] = ""
 	}
 
+	matches["name"] = name
```
Comment (wking) [inline tools/hostbasedimagenames/hostbasedimagenames.go]: This needs a corresponding function-comment update.

File: tools/indextemplate/indextemplate.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	request := &http.Request{
+		Method: "GET",
+		URL:    uri,
+		Header: map[string][]string{
+			"Accept": {"application/vnd.oci.image.index.v1+json"},
+		},
+	}
+	request = request.WithContext(ctx)
+
+	logrus.Debugf("requesting an application/vnd.oci.image.index.v1+json from %s", request.URL)
```
Comment (wking) [inline tools/indextemplate/indextemplate.go]: DRY up by pulling media type from `request`.

File: tools/newimagespec/newimagespec.go (hunk 1 above)
Comment (wking) [reply tools/newimagespec/newimagespec.go]: > Can this be defined as below?

Done with 5faa7c7 → 44b271f.

File: tools/hostbasedimagenames/hostbasedimagenames.go (hunk 2 above)
Comment (wking) [reply tools/hostbasedimagenames/hostbasedimagenames.go]: > This needs a corresponding function-comment update.

Done with 5faa7c7 → 44b271f.

File: tools/indextemplate/indextemplate.go (hunk 3 above)
Comment (wking) [reply tools/indextemplate/indextemplate.go]: > DRY up by pulling media type from `request`.

Done with 5faa7c7 → 44b271f.

**Comment by xiekeyang** [comment]
LGTM

**Pull request merged**